compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1516,6 +1518,15 @@ impl<'a> Parser<'a> {
             || self.token.is_whole_block()
         {
             self.parse_block_expr(label, lo, BlockCheckMode::Default)
+        } else if !ate_colon
+            && (matches!(self.token.kind, token::CloseDelim(_) | token::Comma)
```

Comment:
I'm using "close delim or comma or operator" here as a heuristic, because if we use something like "next token isn't colon" then we suggest `'a 0` => `'a' 0` which is nonsense.